### PR TITLE
[FLOC-3482] Retry node creation failure and cleanup node resources

### PR DESCRIPTION
**Pull request by exarkun**
Fixes https://clusterhq.atlassian.net/browse/FLOC-3482

### CODE REVIEW

File: flocker/provision/_libcloud.py [hunk 1]
```diff
@@ -222,3 +229,150 @@ def create_node(self, name, distribution,
...
             distribution=distribution)
+
+    def _create_with_retry(self, **kwargs):
+        """
+        Create a compute instance.  Clean up failed attempts and retry if
+        necessary.
+        """
+        return poll_until(lambda: self._get_node(**kwargs), iter([15] * 10))
+
+    def _get_node(self, name, **kwargs):
```
Comment (jml) [inline flocker/provision/_libcloud.py]: Why 'get' and not 'create'?

File: flocker/provision/_libcloud.py [hunk 2]
```diff
@@ -222,3 +229,150 @@ def create_node(self, name, distribution,
...
+        }
+
+        try:
+            poll_until(
+                predicate=lambda: self._node_in_state(node, terminal_states),
+                # Overall retry sleep time (not quite the same as timeout since
+                # it doesn't count time spent in the predicate) is just copied
+                # from the default libcloud timeout for wait_until_running.
+                # Maybe some other value would be better.
+                steps=iter([15] * (600 / 15)),
```
Comment (jml) [inline flocker/provision/_libcloud.py]: No action required.

This suggests two improvements to our shared retry logic:

 1. A 'steps' generator that is aware of how much time has passed in the poll and reduces wait time  accordingly
 2. Changing the interface such that the steps generator yields timedeltas

File: flocker/provision/_libcloud.py [hunk 3]
```diff
@@ -222,3 +229,150 @@ def create_node(self, name, distribution,
...
+        nodes = _retry_exception(self._driver.list_nodes)
+        for node in nodes:
+            if node.uuid == target_node.uuid:
+                return node.state in target_states
+        return False
+
+    def _cleanup_node_named(self, name):
+        """
+        Destroy a node with the given name, if there is one.
+        """
```
Comment (jml) [inline flocker/provision/_libcloud.py]: Please add, "Otherwise, do nothing."

**Comment by jml** [comment]
Wow. Thanks for getting to the bottom of this. It's good to have a more stable foundation to build on.

File: flocker/provision/_libcloud.py (hunk 1 above)
Comment (exarkun) [inline flocker/provision/_libcloud.py]: They're synonyms right? :wink:

**Pull request merged**